app/src/test/java/de/westnordost/streetcomplete/data/osm/mapdata/MapDataCacheTest.kt

```diff
@@ -735,4 +735,85 @@ internal class MapDataCacheTest {
         assertTrue(cache.getRelationsForNode(3L) { relationDB.getAllForNode(it) }.containsExactlyInAnyOrder(listOf(rel2)))
         verify(relationDB).getAllForNode(3L) // was fetched from cache
     }
+
+    @Test fun `getMapDataWithGeometry fetches all and caches if nothing is cached`() {
+        val node1 = node(1, LatLon(0.0, 0.0)) // TilePos(x=32768, y=32768)
+        val node2 = node(2, LatLon(0.0001, 0.0001)) // TilePos(x=32768, y=32767)
+        val node3 = node(3, LatLon(-0.0001, 0.0001)) // TilePos(x=32768, y=32768)
+        val nodesBBox = listOf(node1.position, node2.position, node3.position).enclosingBoundingBox()
+        val nodesRect = nodesBBox.enclosingTilesRect(16)
+        assertTrue(nodesRect.size == 2)
+        val way1 = way(1, nodes = listOf(1L, 2L))
+        val way2 = way(2, nodes = listOf(3L, 1L))
+        val way3 = way(3, nodes = listOf(3L, 2L))
+        val rel1 = rel(1, members = listOf(RelationMember(ElementType.NODE, 1L, "")))
+        val rel2 = rel(2, members = listOf(RelationMember(ElementType.WAY, 1L, "")))
+        val elementsInsideBBox = listOf(node1, node2, node3, way1, way2, way3, rel1, rel2)
+        val elementDB: ElementDao = mock()
+        on(elementDB.getAll(nodesRect.asBoundingBox(16))).thenReturn(elementsInsideBBox).thenThrow(IllegalStateException())
+        val cache = MapDataCache(16, 4, 10) { elementDB.getAll(it) to emptyList() }
+
+        val expectedMapData = MutableMapDataWithGeometry().apply {
+            listOf(node1, node2, node3).forEach { put(it, ElementPointGeometry(it.position)) }
+            listOf(way1, way2, way3, rel1, rel2).forEach { put(it, null) }
+            boundingBox = nodesBBox
+        }
+        assertEquals(expectedMapData, cache.getMapDataWithGeometry(nodesBBox))
+        verify(elementDB).getAll(nodesRect.asBoundingBox(16))
+        // second time it's cached
+        assertEquals(expectedMapData, cache.getMapDataWithGeometry(nodesBBox))
```

Comment:
Wouldn't you have to test for no further interactions with elementDB here? (Or is this implied in mockito - not sure)

Reply:
Not sure how I can test that...
What I did is line 753:
`on(elementDB.getAll(nodesRect.asBoundingBox(16))).thenReturn(elementsInsideBBox).thenThrow(IllegalStateException())`
so on second request there is an exception. You can check this be adding `cache.clear()` before the last `assertEquals`.

Reply:
Ah I see, that should work, I didn't notice that.

Maybe it would be better readable to use `verifyNoMoreInteractions``?

Reply:
I didn't know about that...